### PR TITLE
Optimize dataset hierarchy search

### PR DESCRIPTION
**Pull request by IvoTod**
## Background
When collecting all datasets, for every single dataset we ran a query to determine whether the user has access to it via hierarchy. The overhead from running many SQL queries became large enough on the production instance to be the second slowest part of loading datasets.

## Aim
Optimize dataset accessibility collection.

## Implementation
A new query was written, which using CTEs would collect all of the user's datasets from the user's groups and join this set with the branches of the hierarchy, This query was implemented into the django permission class `IsDatasetAllowed` and into it's `permitted_datasets` static method. The clients of IsDatasetAllowed were also updated to use this new query.

### CODE REVIEW

File: wdae/wdae/datasets_api/permissions.py [hunk 1]
```diff
@@ -43,29 +48,174 @@ def has_permission(self, request: HttpRequest, view: Any) -> bool:
...
+        ),
+        --
+        dataset_to_group AS (
+            SELECT
+                d.id AS did,
+                d.dataset_id AS dname,
+                g.id AS gid,
+                g.name AS gname
+            FROM
+                datasets_api_dataset d
```
Comment (iordanivanov) [inline wdae/wdae/datasets_api/permissions.py:99]: AS d?

File: wdae/wdae/datasets_api/permissions.py [hunk 2]
```diff
@@ -43,29 +48,174 @@ def has_permission(self, request: HttpRequest, view: Any) -> bool:
...
+                LEFT OUTER JOIN datasets_api_dataset_groups dg
```
Comment (iordanivanov) [inline wdae/wdae/datasets_api/permissions.py:100]: AS dg?

File: wdae/wdae/datasets_api/permissions.py [hunk 3]
```diff
@@ -43,29 +48,174 @@ def has_permission(self, request: HttpRequest, view: Any) -> bool:
...
+        ) AS (
+            SELECT
+                t.id AS dataset_id,
+                t.id AS descendant_id,
+                t.dataset_id AS descendant_wdae_id,
+                h.instance_id as instance_id,
+                0
+            FROM
+                datasets_api_dataset t
+            LEFT OUTER JOIN datasets_api_datasethierarchy h
```
Comment (iordanivanov) [inline wdae/wdae/datasets_api/permissions.py:117]: AS t?
AS h?

File: wdae/wdae/datasets_api/permissions.py [hunk 4]
```diff
@@ -43,29 +48,174 @@ def has_permission(self, request: HttpRequest, view: Any) -> bool:
...
+        ) AS (
+            SELECT
+                t.id AS dataset_id,
+                t.id AS ancestor_id,
+                t.dataset_id AS ancestor_wdae_id,
+                h.instance_id as instance_id,
+                0
+            FROM
+                datasets_api_dataset t
+            LEFT OUTER JOIN datasets_api_datasethierarchy h
```
Comment (iordanivanov) [inline wdae/wdae/datasets_api/permissions.py:150]: AS t?
AS h?

File: wdae/wdae/datasets_api/permissions.py [hunk 5]
```diff
@@ -43,29 +48,174 @@ def has_permission(self, request: HttpRequest, view: Any) -> bool:
...
+        ),
+        to_root (
+            dataset_id, ancestor_id, ancestor_wdae_id,
+            instance_id, DEPTH
+        ) AS (
+            SELECT
+                t.id AS dataset_id,
+                t.id AS ancestor_id,
+                t.dataset_id AS ancestor_wdae_id,
+                h.instance_id as instance_id,
```
Comment (iordanivanov) [inline wdae/wdae/datasets_api/permissions.py:146]: capital AS?

File: wdae/wdae/datasets_api/permissions.py [hunk 6]
```diff
@@ -43,29 +48,174 @@ def has_permission(self, request: HttpRequest, view: Any) -> bool:
...
+                ON h.ancestor_id = t.id and h.descendant_id = t.id
+            UNION ALL
+            SELECT
+                t.dataset_id,
+                h.ancestor_id,
+                hd.dataset_id,
+                h.instance_id,
+                t.DEPTH - 1
+            FROM
+                to_root t
```
Comment (iordanivanov) [inline wdae/wdae/datasets_api/permissions.py:160]: AS t?

File: wdae/wdae/datasets_api/permissions.py [hunk 7]
```diff
@@ -43,29 +48,174 @@ def has_permission(self, request: HttpRequest, view: Any) -> bool:
...
+                from_root.descendant_id,
+                from_root.descendant_wdae_id,
+                from_root.instance_id
+            FROM
+                from_root
+        )
+        SELECT DISTINCT db.branch_dataset_id, db.branch_dataset_wdae_id
+        FROM user_to_group ug
+        LEFT OUTER JOIN dataset_to_group dg ON ug.gid = dg.gid
+        LEFT OUTER JOIN dataset_branch db ON db.dataset_id = dg.did
```
Comment (iordanivanov) [inline wdae/wdae/datasets_api/permissions.py:195]: AS ug?
AS dg?
AS db?

**Review by iordanivanov — APPROVED** [review]

File: wdae/wdae/datasets_api/permissions.py [hunk 8]
```diff
@@ -43,29 +48,174 @@ def has_permission(self, request: HttpRequest, view: Any) -> bool:
...
-        if user_has_permission(
+        return user_has_permission(
             wgpf_instance.instance_id, cast(User, request.user), obj,
-        ):
-            return True
+        )
 
-        return False
+    @staticmethod
+    def get_allowed_datasets_query() -> str:
```
Comment (lchorbadjiev) [inline wdae/wdae/datasets_api/permissions.py:59]: `get_allowed_` is too broad... consider using `construct` or `build` or `prepare` instead of `get`

**Review by lchorbadjiev — APPROVED** [review]

**Pull request merged**